addons/l10n_it_edi/models/account_move.py

```diff
@@ -778,12 +778,14 @@ def _l10n_it_edi_create_move_with_attachment(self, filename, content, key, proxy
 
         return move
 
-    def _l10n_it_edi_search_partner(self, company, vat, codice_fiscale, email):
+    def _l10n_it_edi_search_partner(self, company, vat, codice_fiscale, email, destination_code=None):
+        base_domain = self.env['res.partner']._check_company_domain(company)
+        if destination_code:
+            base_domain += [('l10n_it_pa_index', 'in', (destination_code, False, ''))]
         for domain in [vat and [('vat', 'ilike', vat)],
                        codice_fiscale and [('l10n_it_codice_fiscale', 'in', ('IT' + codice_fiscale, codice_fiscale))],
                        email and ['|', ('email', '=', email), ('l10n_it_pec_email', '=', email)]]:
-            if domain and (partner := self.env['res.partner'].search(
-                    domain + self.env['res.partner']._check_company_domain(company), limit=1)):
+            if domain and (partner := self.env['res.partner'].search(domain + base_domain, limit=1, order='l10n_it_pa_index desc')):
```

Comment:
Woah, I wouldn't add an `AND` condition in stable, it would be a large change of behavior.
Suppose someone has put "X" in the CodiceDestinatario for any reason: the match would start failing.

```suggestion
    def _l10n_it_edi_search_partner(self, company, vat, codice_fiscale, email, destination_code=None):
        for domain in [vat and destination_code 
                           and [('vat', 'ilike', vat), ('l10n_it_pa_index', 'ilike', destination_code)],
                       vat and [('vat', 'ilike', vat)],
                       codice_fiscale and [('l10n_it_codice_fiscale', 'in', ('IT' + codice_fiscale, codice_fiscale))],
                       email and ['|', ('email', '=', email), ('l10n_it_pec_email', '=', email)]]:
            if domain and (partner := self.env['res.partner'].search(domain, limit=1)):
```

Reply:
I agree, but this proposal would only check the destination code in combination with the VAT. Wouldn’t it also be relevant to check for the codice fiscale and the email as well?

Perhaps with the following priority order:
	•	VAT and Destination
	•	Codice Fiscale and Destination
	•	Email and Destination
	•	VAT only
	•	Codice Fiscale only
	•	Email only

A similar fix could also be implemented in version 16. I didn’t do it initially because this is more of a “bugprovement,” and version 16 uses a different method (_l10n_it_get_partner_invoice) compared to the newer versions (17+). Should I proceed with a backport?

Reply:
Maybe we can add a single query that does search for
(
    (vat and pa_index)
 or (codice_fiscale and pa_index)
)
... or just search for  ( pa_index ) itself 🤣

Reply:
Yes, I also received the email stating that only critical issues should be fixed in stable.
Would you like me to rebase it onto master?

We can’t perform a direct search on the pa_index; is this value not unique?

What do you think about one of these two proposals? The first one only adds a single query:
```python
# Solution 1
def _l10n_it_edi_search_partner(self, company, vat, codice_fiscale, email, destination_code=None):
    base_domain = self.env['res.partner']._check_company_domain(company)
    vat_expression = ('vat', 'ilike', vat)
    codice_fiscale_expression = ('l10n_it_codice_fiscale', 'in', ('IT' + codice_fiscale, codice_fiscale))
    for domain in [destination_code and (vat or codice_fiscale)
                   and [('l10n_it_pa_index', 'ilike', destination_code), vat_expression if vat else codice_fiscale_expression],
                   vat and [vat_expression],
                   codice_fiscale and [codice_fiscale_expression],
                   email and ['|', ('email', '=', email), ('l10n_it_pec_email', '=', email)]]:
        if domain and (partner := self.env['res.partner'].search(domain + base_domain, limit=1)):
            return partner
    return self.env['res.partner']

# Solution 2
def _l10n_it_edi_search_partner(self, company, vat, codice_fiscale, email, destination_code=None):
    base_domain = self.env['res.partner']._check_company_domain(company)
    pa_index_domain = [('l10n_it_pa_index', 'ilike', destination_code)]
    vat_domain = [('vat', 'ilike', vat)]
    codice_domain = [('l10n_it_codice_fiscale', 'in', ('IT' + codice_fiscale, codice_fiscale))]
    for domain in [vat and destination_code and expression.AND([vat_domain, pa_index_domain]),
                   codice_fiscale and destination_code and expression.AND([codice_domain, pa_index_domain]),
                   vat and vat_domain,
                   codice_fiscale and codice_domain,
                   email and ['|', ('email', '=', email), ('l10n_it_pec_email', '=', email)]]:
        if domain and (partner := self.env['res.partner'].search(domain + base_domain, limit=1)):
            return partner
    return self.env['res.partner']

```